### PR TITLE
Implement EMCC_FORCE_STDLIBS using --whole-archive

### PR DESCRIPTION
**Pull request by sbc100**
Previously when using EMCC_FORCE_STDLIBS we were building a second
`.bc` copy of any `.a` libraries.  Not that we have
--whole-archive/--no-whole-archive we no longer need to do this.

This change also refactors a bunch of code in system_libs that
includes a couple of fixes:

- Libraries can now depend in other libraries that appear earlier in
  the list.
- Libraries that are depended on by forced libraries are not themselves
  treated as forced.

### CODE REVIEW

File: tools/system_libs.py [hunk 1]
```diff
@@ -574,99 +579,126 @@ class Dummy(object):
...
+      add_library(system_libs_map[d])
+
+  # Go over libraries to figure out which we must include
+  for lib in system_libs:
+    if lib.shortname in included:
+      continue
+    force_this = lib.shortname in force
+    if not force_this and only_forced:
+      continue
+    include_this = force_this or lib.shortname in include
```
Comment (jgravelle-google) [inline tools/system_libs.py]: Wait we have an `include` and an `included`? That's really hard to keep straight at first glance. We should rename one of them to `included_in_link` or `to_be_included` or `to_include` or something to make it more than a one-char difference

File: tools/system_libs.py [hunk 2]
```diff
@@ -574,99 +579,126 @@ class Dummy(object):
...
-
-  for actual in ret:
-    if os.path.basename(actual) == 'libcxxabi.bc':
-      # libcxxabi and libcxx *static* linking is tricky. e.g. cxa_demangle.cpp disables c++
-      # exceptions, but since the string methods in the headers are *weakly* linked, then
-      # we might have exception-supporting versions of them from elsewhere, and if libcxxabi
-      # is first then it would "win", breaking exception throwing from those string
-      # header methods. To avoid that, we link libcxxabi last.
-      ret = [f for f in ret if f != actual] + [actual]
+    ret.append((shared.Cache.get('wasm_compiler_rt.a', lambda: create_wasm_compiler_rt('wasm_compiler_rt.a'), extension='a'), False))
```
Comment (jgravelle-google) [inline tools/system_libs.py]: I'm not fond of the tuple-ing of this, but I can't think of a better way to do it. :/

File: tools/system_libs.py [hunk 3]
```diff
@@ -574,99 +579,126 @@ class Dummy(object):
...
+    name = shortname + '.' + lib.suffix
+
+    logging.debug('including %s' % name)
+
+    def do_create():
+      return lib.create(name)
+
+    libfile = shared.Cache.get(name, do_create, extension=lib.suffix)
+    need_whole_archive = lib.shortname in force and lib.suffix != 'bc'
+    ret.append((libfile, need_whole_archive))
```
Comment (jgravelle-google) [inline tools/system_libs.py]: It's kinda weird that this is called ret, but is closed over. As in, we have a ret variable, that we do not return immediately. This implies the name is too short for its scope, on account of it's crossing scopes. We also no longer return ret, we return final.
libfile_pairs?

**Comment by sbc100** [comment]
PTAL

File: tools/system_libs.py [hunk 4]
```diff
@@ -502,9 +503,13 @@ def create_wasm_libc_rt(libname):
...
-  force = set((force.split(',') if force else []) + forced)
-  if force:
-    logging.debug('forcing stdlibs: ' + str(force))
+  force_include = set((force.split(',') if force else []) + forced)
+  if force_include:
+    logging.debug('forcing stdlibs: ' + str(force_include))
+
+  # Set of libraries to include on the link line, as opposed to `force` which
+  # is the set of libraries to force include (with --whole-archive).
+  alwasy_include = set()
```
Comment (jgravelle-google) [inline tools/system_libs.py]: typo: `always_include`

Comment (sbc100) [reply tools/system_libs.py]: My brain really can't tell the difference.., sigh.

**Pull request merged**